docs/spec/lightclient/Lightclient_A_1.tla

```diff
@@ -0,0 +1,463 @@
+        \* the latest verified block the the trusted block
+        /\ latestVerified = trustedLightBlock
+
+\* copy the block from the light store        
+LightStoreGetInto(block, height) ==
+    block = fetchedLightBlocks[height]
+
+\* block should contain a copy of the block from the reference chain, with a matching commit
+CopyLightBlockFromChain(block, height) ==
+    LET ref == blockchain[height]
+        lastCommit ==
+          IF height < ULTIMATE_HEIGHT
+          THEN blockchain[height + 1].lastCommit
+            \* for the ultimate block, which we never use, as ULTIMATE_HEIGHT = TARGET_HEIGHT + 1
+          ELSE blockchain[height].VS 
+    IN
+    block = [header |-> ref, Commits |-> lastCommit]      
+
+\* Either the primary is correct and the block comes from the reference chain,
+\* or the block is produced by a faulty primary.
+\*
+\* TODO: add a traceability tag when Josef adds it in the English spec
+FetchLightBlockInto(block, height) ==
+    IF IS_PRIMARY_CORRECT
+    THEN CopyLightBlockFromChain(block, height)
+    ELSE BC!IsLightBlockAllowedByDigitalSignatures(height, block)
+
+\* add a block into the light store    
+\*
+\* TODO: add a traceability tag when Josef adds it in the English spec
+LightStoreUpdateBlocks(lightBlocks, block) ==
+    LET ht == block.header.height IN    
+    [h \in DOMAIN lightBlocks \union {ht} |->
+        IF h = ht THEN block ELSE lightBlocks[h]]
+
+\* update the state of a light block      
+\*
+\* TODO: add a traceability tag when Josef adds it in the English spec
+LightStoreUpdateStates(statuses, ht, blockState) ==
+    [h \in DOMAIN statuses \union {ht} |->
+        IF h = ht THEN blockState ELSE statuses[h]]      
+
+\* Check, whether newHeight is a possible next height for the light client.
+\*
+\* TODO: add a traceability tag when Josef adds it in the English spec
+ScheduleTo(newHeight, pNextHeight, pTargetHeight, pLatestVerified) ==
+    LET ht == pLatestVerified.header.height IN
+    \/ /\ ht = pNextHeight
+       /\ ht < pTargetHeight
+       /\ pNextHeight < newHeight
+       /\ newHeight <= pTargetHeight
+    \/ /\ ht < pNextHeight
+       /\ ht < pTargetHeight
+       /\ ht < newHeight
+       /\ newHeight < pNextHeight
+    \/ /\ ht = pTargetHeight
+       /\ newHeight = pTargetHeight
+
+\* The loop of VerifyToTarget.
+\*
+\* TODO: add a traceability tag when Josef adds it in the English spec
+VerifyToTargetLoop ==
+      \* the loop condition is true
+    /\ latestVerified.header.height < TARGET_HEIGHT
+      \* pick a light block, which will be constrained later
+    /\ \E current \in BC!LightBlocks:
+        \* Get next LightBlock for verification
+        /\ IF nextHeight \in DOMAIN fetchedLightBlocks
+           THEN /\ LightStoreGetInto(current, nextHeight)
+                /\ UNCHANGED fetchedLightBlocks
+           ELSE /\ FetchLightBlockInto(current, nextHeight)
+                /\ fetchedLightBlocks' = LightStoreUpdateBlocks(fetchedLightBlocks, current)
```

Comment:
Would it be possible to make this into something like:
```suggestion
        /\ IF GetFromStore(current, store, nextHeight) THEN
                newStore = store
           ELSE 
                /\ FetchLightBlockInto(current, nextHeight)
                /\ newStore = UpdateStore(store, current)
```
...so it's easier to follow with the spec and code.

Reply:
This will be hard to do without introducing side effects. In your suggestion, `current` and `store` are updated simultaneously. In the proposed code, we would compute new `store` and constrain `current`. These are two different things that should not happen in the same place, for clarity of the spec.

It would be really cool to discuss this question and the above question in a dev session. This is really something, where the programming ends and logic starts :-)